### PR TITLE
Upgrade wasm crate dependencies

### PR DESCRIPTION
**Pull request by athei**
This upgrades:

```
parity-wasm 0.42 -> 0.45
wasm-instrument 0.1 -> 0.2
wasmi 0.9 -> 0.13
```

These do not contain big changes but merely big fixes. wasmi is upgraded to the last version before bigger changes were made. We will upgrade only until we did more testing: https://github.com/paritytech/roadmap/issues/9

We should also do a burn in with wasmi as a execution engine.

### CODE REVIEW

**Review by koute — APPROVED** [review]
LGTM.

Performance-wise just for your reference, I have a few benchmarks laying around that I've created recently (when investigating the random performance regression on wasmtime bump); here they are:

```
coremark,wasmi_009_under_wasmtime_040,217
coremark,wasmi_013_under_wasmtime_040,203
coremark,wasmi_016_under_wasmtime_040,180
regexredux,wasmi_009_under_wasmtime_040,2280
regexredux,wasmi_013_under_wasmtime_040,2386
regexredux,wasmi_016_under_wasmtime_040,2645
noop,wasmi_009_under_wasmtime_040,9237
noop,wasmi_013_under_wasmtime_040,9208
noop,wasmi_016_under_wasmtime_040,4352
```

First column is the name of the benchmark, the second column is the executor, and the third is the score or time. For `coremark` higher is better, for `regexredux` and `noop` lower is better. It seems that (at least in these benchmarks) the higher you bump wasmi the worse the performance gets in CPU-heavy benchmarks (`coremark` and `regexredux`) while the performance in an instantiation-heavy benchmark `noop` gets better on 0.16.

**Review by pepyakin — APPROVED** [review]

**Comment by athei** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_contracts

**Comment by Robbepop** [comment]
> LGTM.
> 
> Performance-wise just for your reference, I have a few benchmarks laying around that I've created recently (when investigating the random performance regression on wasmtime bump); here they are:
> 
> ```
> coremark,wasmi_009_under_wasmtime_040,217
> coremark,wasmi_013_under_wasmtime_040,203
> coremark,wasmi_016_under_wasmtime_040,180
> regexredux,wasmi_009_under_wasmtime_040,2280
> regexredux,wasmi_013_under_wasmtime_040,2386
> regexredux,wasmi_016_under_wasmtime_040,2645
> noop,wasmi_009_under_wasmtime_040,9237
> noop,wasmi_013_under_wasmtime_040,9208
> noop,wasmi_016_under_wasmtime_040,4352
> ```
> 
> First column is the name of the benchmark, the second column is the executor, and the third is the score or time. For `coremark` higher is better, for `regexredux` and `noop` lower is better. It seems that (at least in these benchmarks) the higher you bump wasmi the worse the performance gets in CPU-heavy benchmarks (`coremark` and `regexredux`) while the performance in an instantiation-heavy benchmark `noop` gets better on 0.16.

Have you used
```
[profile.release]
lto = "fat"
codegen-units = 1
```
while profiling? Especially for newer `wasmi` versions this makes a huge performance difference (>200%).

**Comment by koute** [comment]
> Have you used
> 
> ```toml
> [profile.release]
> lto = "fat"
> codegen-units = 1
> ```
> 
> for `wasmi` compilation while profiling? Especially for newer `wasmi` versions this makes a huge performance difference (>200%). More information here: [paritytech/wasmi#339](https://github.com/paritytech/wasmi/issues/339)

Yes.

All of the WASM kernels were compiled on the following profile:

```
[profile.lto]
inherits = "release"
lto = true
codegen-units = 1
```

(AFAIK `true` is same as `"fat"`)

**Comment by Robbepop** [comment]
> 

Is there a way to check this? I often had problem properly setting up the profile for dependencies. The thing is that `wasmi` 0.16.0 should be vastly more efficient than version 0.9.0 and 0.13.0 but this benchmark shows quite the opposite which is only true if the wrong profile was selected since the newer `wasmi` version is quite a lot slower with the wrong profiles.

**Comment by athei** [comment]
@Robbepop you conducted your benchmarks only with wasmi compiled to native, right? These benchmarks run it under wasmtime which could completely change the game.

**Comment by Robbepop** [comment]
> @Robbepop you conducted your benchmarks only with wasmi compiled to native, right? These benchmarks run it under wasmtime which could completely change the game.

I am aware that Wasm compilation might change things but I just doubt that the difference between Wasm and native compilation run on a JIT compiler is as big as claimed in the benchmarks above. That's it. I'd like to have sources to those benchmarks to convince myself.

**Comment by athei** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_contracts

**Comment by koute** [comment]
> @koute Is there a way to check this?

Not that I know of, although you could probably do something like this in `build.rs` to check that the profile matches:

```
pub fn main() {
    assert_eq!(std::env::var("PROFILE"), Some("lto"));
}
```

> Maybe forgot to set the `lto` profile? An easy way to check this is to compare performance when benchmarking explicitly with the standard `release` profile which should yield drastic performance differences for `wasmi` `0.16.0`.

I might have screwed up something else, but this I shouldn't have. (:

Here are the results for the standard `release` profile (the underlying kernels which `wasmi` runs are still compiled with `lto`) in comparison with the previous numbers and also some numbers without `wasmtime`:

| benchmark | runtime | release | lto |
| --------- | ------- | ------- | --- |
|coremark|wasmi_009_under_wasmtime_040|248|217|
|coremark|wasmi_013_under_wasmtime_040|263|203|
|coremark|wasmi_016_under_wasmtime_040|170|180|
|coremark|wasmi_009_on_bare_metal|452|373|
|coremark|wasmi_013_on_bare_metal|393|411|
|coremark|wasmi_016_on_bare_metal|253|581|
|regexredux|wasmi_009_under_wasmtime_040|2022|2280|
|regexredux|wasmi_013_under_wasmtime_040|2022|2386|
|regexredux|wasmi_016_under_wasmtime_040|2816|2645|
|regexredux|wasmi_009_on_bare_metal|1269|1353|
|regexredux|wasmi_013_on_bare_metal|1333|1241|
|regexredux|wasmi_016_on_bare_metal|1902|891|
|noop|wasmi_009_under_wasmtime_040|10562|9237|
|noop|wasmi_013_under_wasmtime_040|9197|9208|
|noop|wasmi_016_under_wasmtime_040|6347|4352|
|noop|wasmi_009_on_bare_metal|1762|1697|
|noop|wasmi_013_on_bare_metal|1791|1823|
|noop|wasmi_016_on_bare_metal|761|748|

**Comment by koute** [comment]
@Robbepop Okay, I've cleaned up my benchmarks just a little bit and pushed them into a repo; you can check them out here if you want: https://github.com/koute/wasm-bench

**Comment by Robbepop** [comment]
@koute Thanks a lot for making the benchmarks available. They provided me with some valuable insights.

I was able to reproduce your results on my own machine where newer `wasmi` versions perform worse than older ones.
However, coming from the ink! project I already knew that Rust+LLVM do quite a bad job at optimizing for Wasm targets.
Therefore I ran Binaryen's `wasm-opt` over all the `wasmi_kernel.wasm` from all the `wasmi` versions and reran the benchmarks.

These are the results I got afterwards:


| Benchmark    | `wasmi` and Wasmtime versions | Result |
|--------------|-------------------------------|--------|
| `regexredux` | wasmi_016_under_wasmtime_040  | 1433 |
| `regexredux` | wasmi_013_under_wasmtime_040  | 2769 |
| `regexredux` | wasmi_009_under_wasmtime_040  | 1947 |
| `regexredux` | wasmi_016_under_wasmtime_036  | 1590 |
| `regexredux` | wasmi_013_under_wasmtime_036  | 2213 |
| `regexredux` | wasmi_009_under_wasmtime_036  | 1795 |
| `regexredux` | wasmi_016_on_bare_metal       | 792 |
| `regexredux` | wasmi_013_on_bare_metal       | 1243 |
| `regexredux` | wasmi_009_on_bare_metal       | 1315 |
| `coremark`   | wasmi_016_under_wasmtime_040  | 364.5422 |
| `coremark`   | wasmi_013_under_wasmtime_040  | 158.76796 |
| `coremark`   | wasmi_009_under_wasmtime_040  | 243.26955 |
| `coremark`   | wasmi_016_under_wasmtime_036  | 311.86652 |
| `coremark`   | wasmi_013_under_wasmtime_036  | 217.20244 |
| `coremark`   | wasmi_009_under_wasmtime_036  | 281.19507 |
| `coremark`   | wasmi_016_on_bare_metal       | 693.30646 |
| `coremark`   | wasmi_013_on_bare_metal       | 433.965 |
| `coremark`   | wasmi_009_on_bare_metal       | 415.36865 |
| `noop`       | wasmi_016_under_wasmtime_040  | 9500 |
| `noop`       | wasmi_013_under_wasmtime_040  | 18258 |
| `noop`       | wasmi_009_under_wasmtime_040  | 18179 |
| `noop`       | wasmi_016_under_wasmtime_036  | 8391 |
| `noop`       | wasmi_013_under_wasmtime_036  | 16593 |
| `noop`       | wasmi_009_under_wasmtime_036  | 16401 |
| `noop`       | wasmi_016_on_bare_metal       | 1525 |
| `noop`       | wasmi_013_on_bare_metal       | 3511 |
| `noop`       | wasmi_009_on_bare_metal       | 2996 |

In conclusion, after `wasm-opt -O4` was applied to the `wasmi_kernel.wasm` then `wasmi` version `0.16.0` performs by far the best under Wasmtime whereas `wasmi` version `0.13.0` performs the worst. However, all `wasmi` versions perform better after being post-optimized by `wasm-opt`.

My recommendation therefore is to use `wasm-opt` for the Substrate runtime if we do not do this already.

**Comment by athei** [comment]
We don't. I have a branch where I tested this. However it is a bit experimental because `wasm-opt` rust wrapper is still experimental. I intend to merge this as soon as this becomes more stable. I don't want require users to install wasm-opt to compile substrate.

**Comment by athei** [comment]
bot merge
